File: charter/dispatch.py
"""Committed dispatch tally — how often each persona is actually *used*.

Roster health used to be judged by memory volume alone, which is blind to the failure
that matters: a persona that exists, advertises a `delegate-when`, lints green, and is
**never dispatched** while the work it owns routes to a generic sub-agent. Memory counts
can't see that; only dispatch counts can.

Shape of the store::

    personas/_dispatch/<YYYY-MM>.<host>.jsonl     one JSON object per line

Three properties matter:

* **Counts and dates only.** A line is ``{"ts": …, "agent": …}`` for a dispatch, or
  ``{"ts": …, "event": "advice"}`` for routing advice having been shown — never the
  prompt, the description, or any tool input. There is no secret surface to scan, and the
  second shape was added under exactly that constraint: it records that the roster
  appeared, never what it appeared about.
* **Parallel-writer safe.** Lines are small and opened ``O_APPEND``, so concurrent
  dispatches (a fan-out of 8 sub-agents) append atomically without a lock.
* **Conflict-free across engineers.** The filename carries the host, so two people
  recording the same month never touch the same file — the committed tally merges by
  addition instead of conflicting.

Generic agents (``general-purpose``, ``Explore``) are tallied too, under their own name:
the ratio of persona-to-generic dispatch is the whole point, and dropping it would hide
the very gap this store exists to expose.
"""
from __future__ import annotations

import json
import os
import re
import socket
from collections import Counter
from datetime import datetime, timedelta, timezone
from pathlib import Path

from . import config, contain

DIR_NAME = "_dispatch"
#: Agents that aren't personas — tallied so the persona-vs-generic ratio stays visible.
GENERIC = ("general-purpose", "Explore", "claude", "Plan")


def _dir() -> Path:
    return config.PERSONAS_DIR / DIR_NAME


def _host() -> str:
    """Short, filename-safe hostname — keeps each engineer on their own file."""
    raw = (socket.gethostname() or "unknown").split(".")[0]
    return re.sub(r"[^A-Za-z0-9_-]", "", raw)[:32] or "unknown"


def _now() -> datetime:
    return datetime.now(timezone.utc)


def path_for(when: datetime | None = None) -> Path:
    when = when or _now()
    return _dir() / f"{when:%Y-%m}.{_host()}.jsonl"


def record(agent: str, when: datetime | None = None) -> Path | None:
    """Append one dispatch. Best-effort: a tally must never break a turn, so any failure
    is swallowed and reported as None."""
    agent = (agent or "").strip()
    if not agent:
        return None
    when = when or _now()
    p = path_for(when)
    if contain.write_refusal(p):
        return None  # a committed link at this fixed name — see contain.write_refusal
    try:
        p.parent.mkdir(parents=True, exist_ok=True)
        line = json.dumps({"ts": when.isoformat(timespec="seconds"), "agent": agent},
                          sort_keys=True) + "\n"
        # O_APPEND: concurrent sub-agent dispatches interleave safely without a lock.
        fd = os.open(p, os.O_WRONLY | os.O_CREAT | os.O_APPEND, 0o644)
        try:
            os.write(fd, line.encode())
        finally:
            os.close(fd)
        return p
    except OSError:
        return None


#: Marks a row that records ROUTING ADVICE being shown rather than a dispatch happening.
#: Same store, same two-fields-only discipline: a timestamp and this word. The pair
#: "advice shown" vs "dispatches that followed" is the only number that can falsify the
#: bet the roster block rests on — that seeing the roster changes routing — and a bet
#: shipped without it repeats the gap this module's docstring was written about.
ADVICE = "advice"


def record_advice(when: datetime | None = None) -> Path | None:
    """Append one 'routing advice was shown' event. Best-effort, like :func:`record`."""
    when = when or _now()
    p = path_for(when)
    if contain.write_refusal(p):
        return None  # a committed link at this fixed name — see contain.write_refusal
    try:
        p.parent.mkdir(parents=True, exist_ok=True)
        line = json.dumps({"event": ADVICE, "ts": when.isoformat(timespec="seconds")},
                          sort_keys=True) + "\n"
        fd = os.open(p, os.O_WRONLY | os.O_CREAT | os.O_APPEND, 0o644)
        try:
            os.write(fd, line.encode())
        finally:
            os.close(fd)
        return p
    except OSError:
        return None


def advice_tally(days: int | None = None) -> int:
    """How many times routing advice was shown, optionally within the last *days*."""
    rows = [o for o in _read_all() if o.get("event") == ADVICE]
    if days is not None:
        cutoff = _now() - timedelta(days=days)
        rows = [o for o in rows if (t := _ts(o)) and t >= cutoff]
    return len(rows)


#: Marks a row where work was handed to a persona that was ALREADY running — a resume,
#: not a new sub-agent. Its own kind because `DISP` answers "times dispatched as a
#: sub-agent", and folding resumes into that inflates the answer to a question nobody
#: asked, in the column people retire personas on.
RESUME = "resume"


def record_resume(agent: str, when: datetime | None = None) -> Path | None:
    """Append one 'more work handed to a running persona' event. Best-effort.

    `posttooluse_dispatch` fires on Task/Agent, so it sees a sub-agent being CREATED and
    nothing after. Continuing that agent is delegation the tally could not see: on the day
    the roster block shipped, one persona cut two releases and ran a sweep, and the sweep —
    a resume — was invisible to the pair of numbers that exist to measure exactly that.
    """
    agent = (agent or "").strip()
    if not agent:
        return None
    when = when or _now()
    p = path_for(when)
    if contain.write_refusal(p):
        return None  # a committed link at this fixed name — see contain.write_refusal
    try:
        p.parent.mkdir(parents=True, exist_ok=True)
        line = json.dumps({"agent": agent, "event": RESUME,
                           "ts": when.isoformat(timespec="seconds")}, sort_keys=True) + "\n"
        fd = os.open(p, os.O_WRONLY | os.O_CREAT | os.O_APPEND, 0o644)
        try:
            os.write(fd, line.encode())
        finally:
            os.close(fd)
        return p
    except OSError:
        return None


def resume_tally(days: int | None = None) -> int:
    rows = [o for o in _read_all() if o.get("event") == RESUME]
    if days is not None:
        cutoff = _now() - timedelta(days=days)
        rows = [o for o in rows if (t := _ts(o)) and t >= cutoff]
    return len(rows)


def first_advice() -> datetime | None:
    """When routing advice was FIRST shown here, or ``None`` if it never has."""
    stamps = [t for o in _read_all() if o.get("event") == ADVICE and (t := _ts(o))]
    return min(stamps) if stamps else None


def handoffs_since_first_advice() -> int:
    """Dispatches recorded at or after the first advice — the only count that can honestly
    sit beside :func:`advice_tally`.

    The pair is meant to read as fired-vs-followed, and a dispatch that happened before the
    roster ever appeared cannot have followed it. Pairing advice against the LIFETIME total
    is how this line came to claim, on the plane that shipped the feature, that five
    dispatches followed one piece of advice — three of them four days older than the
    feature. Same failure ADR 0016 forbids elsewhere: a conclusion stated with more
    confidence than its provenance earns, here in the very line that exists to check
    whether ADR 0016's mechanism works.

    Still not proof of causation, and the report says "since" rather than "because" for
    that reason. It is a window in which the claim is at least possible.

    Boundary, stated rather than hidden: rows are stamped to the second, so a handoff in
    the SAME second as the first advice counts as after it. It can only over-count, by at
    most the handoffs sharing one second with the moment advice first appeared.
    """
    since = first_advice()
    if since is None:
        return 0
    # Dispatches AND resumes: the pair measures whether the roster moved work to a persona,
    # and it does not care whether that persona was already running.
    return sum(1 for o in _read_all()
               if o.get("agent") and (t := _ts(o)) and t >= since)


#: Month files already parsed, keyed by path → ``((mtime_ns, size), rows)``.
#: Read and written only by :func:`_rows_of`, whose docstring is the argument for it.
_PARSED: dict[str, tuple[tuple[int, int], list[dict]]] = {}


def _rows_of(f: Path) -> list[dict]:
    """The rows of one month file, parsed at most once per version of that file.

    :func:`_read_all` walks EVERY month file ever written, and :func:`tally` sits on a
    per-turn repaint path: `persona.by_use` calls it for the switcher's order,
    `statusline._persona_chip_cells` draws that column, and `frame/panel.run` holds a
    process repainting it for as long as the frame lives. The cost of that walk is
    monotonic in the AGE of the plane and in nothing an operator does — measured at this
    plane's ~225 dispatches a month: 0.37 ms today, 2.10 ms after a year, 10.71 ms after
    five. It degrades silently, and only on the planes that have been used longest, whose
    operators would least expect it (#887).

    A month file that is not the current one is **closed**: :func:`path_for` only ever
    writes ``<this month>.<host>.jsonl``, and nothing reopens an earlier one. So every
    month but one is a permanent hit here, and the walk is bounded by the current month
    however old the plane is.

    **The key is the invalidation, not a summary.** Nothing derived from a row is kept —
    only the rows the jsonl already holds, discarded the instant the file they came from
    is not the file that was read. That is what stops this becoming a second source of
    truth for something the log answers, and ``(path, mtime, size)`` is the same shape
    `statusline._usage_stamp` and `frame/gather` already invalidate on.

    **Size is in the key because mtime alone is a filesystem's promise, not a fact.**
    APFS keeps nanoseconds, but some ext4 configurations report whole seconds, and
    :func:`record` appends to the current month file many times inside one such tick. Two
    versions of an append-only log always differ in length, so the size catches precisely
    what a coarse clock hides.

    **The stat is taken BEFORE the read, and that order is the correctness.** A row
    appended between the two is then memoised under the OLDER stamp and the next call
    re-reads it; taken after, that same row would be memoised under the NEWER stamp and
    stay invisible for the life of the process.

    Rows are shared rather than copied — every caller in this module only reads them, and
    a copy per call would hand back most of what the memo saves. A caller that wants to
    mutate a row must copy it first.
    """
    key = str(f)
    st = f.stat()
    stamp = (st.st_mtime_ns, st.st_size)
    hit = _PARSED.get(key)
    if hit is not None and hit[0] == stamp:
        return hit[1]
    rows: list[dict] = []
    for ln in f.read_text(errors="replace").splitlines():
        try:
            # No blank-line guard above this: `json.loads` raises ValueError on an empty
            # or whitespace-only line exactly as it does on a truncated one, so the pair
            # of lines that used to check for it here was a branch nothing could go red
            # without. It tolerates surrounding whitespace on a real row, too, and
            # `splitlines` has already taken the line endings off.
            o = json.loads(ln)
        except ValueError:
            continue
        if isinstance(o, dict) and (o.get("agent") or o.get("event")):
            rows.append(o)
    _PARSED[key] = (stamp, rows)
    return rows


def _read_all() -> list[dict]:
    d = _dir()
    if not d.exists():
        return []
    out = []
    for f in sorted(d.glob("*.jsonl")):
        try:
            out.extend(_rows_of(f))
        except OSError:
            continue
    return out


def _ts(o: dict) -> datetime | None:
    try:
        t = datetime.fromisoformat(str(o.get("ts", "")))
        return t if t.tzinfo else t.replace(tzinfo=timezone.utc)
    except ValueError:
        return None


def tally(days: int | None = None) -> Counter:
    """agent → dispatch count, optionally limited to the last *days*."""
    rows = _read_all()
    if days is not None:
        cutoff = _now() - timedelta(days=days)
        rows = [o for o in rows if (t := _ts(o)) and t >= cutoff]
    # A resume row carries an agent too, and must NOT land here: this column is read as
    # "times dispatched as a sub-agent", and it is what personas get retired on.
    return Counter(o["agent"] for o in rows if o.get("agent") and not o.get("event"))


def last_seen(agent: str) -> str | None:
    """ISO date this persona was last WORKED, or None. Counts resumes as well as
    dispatches — it feeds the roster block's "last dispatched" line, and a persona that was
    handed work an hour ago has not been idle since whenever it was first created."""
    best = None
    for o in _read_all():
        if o.get("agent") != agent:
            continue
        t = _ts(o)
        if t and (best is None or t > best):
            best = t
    return f"{best:%Y-%m-%d}" if best else None


def generic_share(days: int | None = None) -> tuple[int, int]:
    """(dispatches to generic agents, total dispatches) — the headline routing ratio."""
    t = tally(days)
    return sum(v for k, v in t.items() if k in GENERIC), sum(t.values())


# --------------------------------------------------------------------------- #
# Backfill — seed the tally from past sessions so the baseline exists TODAY     #
# instead of a week from now. Transcripts already record every dispatch; this   #
# just reads what is there. Writes to `*.backfill.jsonl`, kept separate from    #
# live records so a re-run replaces its own output instead of double-counting.  #
# --------------------------------------------------------------------------- #
BACKFILL_SUFFIX = ".backfill.jsonl"


def _transcript_dir() -> Path:
    """Claude Code stores a project's transcripts under ~/.claude/projects/<slug>, where
    the slug is the project path with every separator replaced by '-'."""
    slug = str(config.ROOT).replace("/", "-").replace("\\", "-")
    return Path.home() / ".claude" / "projects" / slug


def _live_keys() -> set[tuple[str, str]]:
    """Every dispatch the live hook actually recorded, by identity — (timestamp, agent).

    This replaced a single global cutoff: the earliest live record, with everything at or
    after it skipped. That prevented double-counting on one assumption — that once live
    recording had started it was **complete** — and #83 is the report that it is not. A
    `PostToolUse(Task|Agent)` hook does not fire for every background dispatch, so three
    real dispatches on 2026-08-10 were missing from a store whose earliest record was
    2026-08-07, and the cutoff put them permanently out of backfill's reach. One early
    live record disabled reconciliation for all later history.

    Identity expresses the same guarantee without the assumption, and it self-heals: a
    window the hook missed is imported the next time backfill runs, whenever that is,
    without anyone having to know which window it was.
    """
    keys: set[tuple[str, str]] = set()
    d = _dir()
    if not d.exists():
        return keys
    for f in d.glob("*.jsonl"):
        if f.name.endswith(BACKFILL_SUFFIX):
            continue
        for ln in f.read_text(errors="replace").splitlines():
            try:
                o = json.loads(ln)
                t = _ts(o)
            except ValueError:
                continue
            if t and o.get("agent"):
                keys.add((t.isoformat(timespec="seconds"), str(o.get("agent"))))
    return keys


def last_backfill() -> datetime | None:
    """When transcripts were last reconciled into the tally, or None if never.

    `stats` reports this because the tally is only as complete as its last reconciliation,
    and a count that silently omits every background dispatch reads exactly like a count
    that includes them (#83)."""
    d = _dir()
    if not d.exists():
        return None
    stamps = [f.stat().st_mtime for f in d.glob(f"*{BACKFILL_SUFFIX}")]
    return datetime.fromtimestamp(max(stamps)) if stamps else None


def scan_transcripts() -> list[dict]:
    """Every sub-agent dispatch recorded in this project's transcripts, as tally rows.
    Reads only the tool name, subagent_type and timestamp — never prompt text."""
    d = _transcript_dir()
    if not d.exists():
        return []
    rows = []
    for f in sorted(d.glob("*.jsonl")):
        try:
            text = f.read_text(errors="replace")
        except OSError:
            continue
        for ln in text.splitlines():
            try:
                o = json.loads(ln)
            except ValueError:
                continue
            if o.get("type") != "assistant":
                continue
            content = (o.get("message") or {}).get("content")
            if not isinstance(content, list):
                continue
            for b in content:
                if not isinstance(b, dict) or b.get("type") != "tool_use":
                    continue
                if b.get("name") not in ("Task", "Agent"):
                    continue
                agent = ((b.get("input") or {}).get("subagent_type") or "").strip()
                if agent:
                    rows.append({"ts": (o.get("timestamp") or "")[:19], "agent": agent})
    return rows


def backfill() -> tuple[int, int]:
    """Seed the tally from transcripts. Returns (imported, skipped-as-already-live).
    Idempotent: rewrites only the backfill files it owns."""
    live = _live_keys()
    keep: dict[str, set[str]] = {}
    seen: set[tuple[str, str]] = set()
    imported = skipped = 0
    for row in scan_transcripts():
        t = _ts(row)
        if not t:
            continue
        key = (t.isoformat(timespec="seconds"), str(row.get("agent") or ""))
        if key in live or key in seen:
            # Already tallied — by the live hook, or by an earlier line of this same run
            # (one dispatch can appear in more than one transcript file).
            skipped += 1
            continue
        seen.add(key)
        name = f"{t:%Y-%m}.{_host()}{BACKFILL_SUFFIX}"
        keep.setdefault(name, set()).add(
            json.dumps({"ts": key[0], "agent": row["agent"]}, sort_keys=True))
        imported += 1
    d = _dir()
    d.mkdir(parents=True, exist_ok=True)
    for f in d.glob(f"*{BACKFILL_SUFFIX}"):  # replace, never append — keeps re-runs clean
        f.unlink()
    for name, lines in keep.items():
        (d / name).write_text("\n".join(sorted(lines)) + "\n")
    return imported, skipped

File: crates/charter-core/src/dispatch.rs
//! The dispatch log: who work was actually routed to. A port of the part of
//! `charter/dispatch.py` the README's persona roster reads — [`tally`] and
//! [`generic_share`].
//!
//! `personas/_dispatch/<month>.<host>.jsonl` is append-only, one JSON object a line, one
//! file per month per machine. It is **committed**, which is what makes the roster block a
//! fact every engineer sees the same way rather than a reading of one laptop.
//!
//! Two rows are written here, both by a hook: [`record`], when a `Task`/`Agent` call returns
//! (`posttooluse-dispatch`), and [`record_resume`], when a `SendMessage` resumes a persona
//! (`posttooluse-message`). Without them the roster `charter docs` draws would count only
//! what the Python charter once logged. Committing the log is not done here: under `share =
//! "commit"` or `"push"` the Python hook commits each row as it lands, and charter-app leaves
//! that to `charter save`, which commits the plane's own files as one decision.

use std::collections::BTreeMap;
use std::path::{Path, PathBuf};

/// The directory under `personas/` the log lives in — `dispatch.DIR_NAME`.
pub const DIR_NAME: &str = "_dispatch";

/// The agent names that are NOT a persona. `dispatch.GENERIC`, in its order.
///
/// A dispatch to one of these is work a persona might have owned, done without it — which
/// is the ratio the roster's headline is about.
pub const GENERIC: [&str; 4] = ["general-purpose", "Explore", "claude", "Plan"];

pub fn dir(root: &Path) -> PathBuf {
    root.join("personas").join(DIR_NAME)
}

/// Every row of every month file, in filename order — `dispatch._read_all`.
///
/// A file that cannot be read is skipped (Python catches `OSError` per file), and so is a
/// line that is not a JSON object carrying an `agent` or an `event`. Bytes that are not
/// UTF-8 are replaced rather than refused, as Python's `read_text(errors="replace")` does:
/// this log is appended to by a hook, and a half-written line must not cost the whole file.
pub fn rows(root: &Path) -> Vec<serde_json::Value> {
    let d = dir(root);
    let Ok(reader) = std::fs::read_dir(&d) else {
        return Vec::new();
    };
    let mut files: Vec<PathBuf> = reader
        .filter_map(Result::ok)
        .map(|e| e.path())
        .filter(|p| p.extension().is_some_and(|e| e == "jsonl"))
        .collect();
    files.sort();
    let mut out = Vec::new();
    for file in files {
        let Ok(bytes) = std::fs::read(&file) else {
            continue;
        };
        let text = String::from_utf8_lossy(&bytes);
        for line in text.lines() {
            let Ok(value) = serde_json::from_str::<serde_json::Value>(line) else {
                continue;
            };
            if !value.is_object() {
                continue;
            }
            if truthy(value.get("agent")) || truthy(value.get("event")) {
                out.push(value);
            }
        }
    }
    out
}

/// Python's truth of a JSON value, for the `o.get("agent") or o.get("event")` guard: a
/// missing key, `null`, `false`, `0` and `""` are all false.
pub(crate) fn truthy(value: Option<&serde_json::Value>) -> bool {
    match value {
        None | Some(serde_json::Value::Null) => false,
        Some(serde_json::Value::Bool(b)) => *b,
        Some(serde_json::Value::String(s)) => !s.is_empty(),
        Some(serde_json::Value::Array(a)) => !a.is_empty(),
        Some(serde_json::Value::Object(o)) => !o.is_empty(),
        Some(serde_json::Value::Number(n)) => n.as_f64() != Some(0.0),
    }
}

/// agent → dispatch count — `dispatch.tally()` with no day window.
///
/// **A resume row carries an agent too, and must not land here.** The column is read as
/// "times dispatched as a sub-agent", and it is what personas get retired on — so a row
/// with an `event` is excluded, whatever else it holds.
///
/// The agent is taken as the string it is. A row whose `agent` is a number is truthy to
/// Python and lands in its `Counter` under that value; here it is skipped, because a name
/// that is not a name matches no persona and could only ever be counted against nothing.
pub fn tally(root: &Path) -> BTreeMap<String, u64> {
    let mut counts: BTreeMap<String, u64> = BTreeMap::new();
    for row in rows(root) {
        if truthy(row.get("event")) {
            continue;
        }
        if let Some(agent) = row.get("agent").and_then(serde_json::Value::as_str)
            && !agent.is_empty()
        {
            *counts.entry(agent.to_string()).or_default() += 1;
        }
    }
    counts
}

/// `(dispatches to generic agents, total dispatches)` — the roster's headline ratio.
pub fn generic_share(counts: &BTreeMap<String, u64>) -> (u64, u64) {
    let generic = counts
        .iter()
        .filter(|(name, _)| GENERIC.contains(&name.as_str()))
        .map(|(_, n)| *n)
        .sum();
    (generic, counts.values().sum())
}

/// This machine's name as a filename part — `dispatch._host`: the first label of the host
/// name, everything outside `[A-Za-z0-9_-]` removed, at most 32 characters, `unknown` for
/// nothing. One file per machine is what keeps two laptops from conflicting on one line.
pub fn host() -> String {
    #[cfg(unix)]
    let raw = rustix::system::uname()
        .nodename()
        .to_string_lossy()
        .into_owned();
    #[cfg(not(unix))]
    let raw = String::new();
    host_of(&raw)
}

/// [`host`]'s rule, for a name already in hand.
pub fn host_of(raw: &str) -> String {
    let first = raw.split('.').next().unwrap_or_default();
    let kept: String = first
        .chars()
        .filter(|c| c.is_ascii_alphanumeric() || matches!(c, '_' | '-'))
        .take(32)
        .collect();
    if kept.is_empty() {
        "unknown".to_string()
    } else {
        kept
    }
}

/// `personas/_dispatch/<YYYY-MM>.<host>.jsonl` for the month `when` falls in (UTC) —
/// `dispatch.path_for`.
pub fn path_for(root: &Path, when: chrono::DateTime<chrono::Utc>, host: &str) -> PathBuf {
    dir(root).join(format!("{}.{host}.jsonl", when.format("%Y-%m")))
}

/// Append one row, keys sorted, as `json.dumps(…, sort_keys=True)` writes it — the three
/// `dispatch.record*` writers' shared body. `None` when containment refuses the path or the
/// write fails.
pub fn append(path: &Path, root: &Path, row: &serde_json::Value) -> Option<PathBuf> {
    use std::io::Write;
    crate::contain::writable(root, path).ok()?;
    std::fs::create_dir_all(path.parent()?).ok()?;
    let mut options = std::fs::OpenOptions::new();
    options.append(true).create(true);
    #[cfg(unix)]
    {
        use std::os::unix::fs::OpenOptionsExt;
        options.mode(0o644);
    }
    let mut file = options.open(path).ok()?;
    let line = format!("{}\n", crate::pyjson::dumps_sorted(row));
    file.write_all(line.as_bytes()).ok()?;
    Some(path.to_path_buf())
}

/// `isoformat(timespec="seconds")` of a UTC instant.
pub fn stamp(when: chrono::DateTime<chrono::Utc>) -> String {
    when.format("%Y-%m-%dT%H:%M:%S+00:00").to_string()
}

/// Log one dispatch to `agent` — `dispatch.record`.
pub fn record(
    root: &Path,
    agent: &str,
    when: chrono::DateTime<chrono::Utc>,
    host: &str,
) -> Option<PathBuf> {
    let agent = crate::memstore::py_strip(agent);
    if agent.is_empty() {
        return None;
    }
    append(
        &path_for(root, when, host),
        root,
        &serde_json::json!({"agent": agent, "ts": stamp(when)}),
    )
}

/// The `event` a resume row carries — `dispatch.RESUME`.
pub const RESUME: &str = "resume";

/// Log that `agent` was resumed rather than dispatched afresh — `dispatch.record_resume`.
pub fn record_resume(
    root: &Path,
    agent: &str,
    when: chrono::DateTime<chrono::Utc>,
    host: &str,
) -> Option<PathBuf> {
    let agent = crate::memstore::py_strip(agent);
    if agent.is_empty() {
        return None;
    }
    append(
        &path_for(root, when, host),
        root,
        &serde_json::json!({"agent": agent, "event": RESUME, "ts": stamp(when)}),
    )
}

#[cfg(test)]
mod tests {
    use super::*;

    fn plane(lines: &[&str]) -> tempfile::TempDir {
        let dir = tempfile::tempdir().unwrap();
        let d = super::dir(dir.path());
        std::fs::create_dir_all(&d).unwrap();
        std::fs::write(d.join("2026-03.host.jsonl"), lines.join("\n")).unwrap();
        dir
    }

    #[test]
    fn a_dispatch_is_counted_and_a_resume_is_not() {
        // A resume row carries an agent; counting it would inflate the column personas are
        // retired on, which is the one thing this tally decides.
        let dir = plane(&[
            r#"{"agent": "devops", "ts": "2026-03-02T09:33:00+00:00"}"#,
            r#"{"agent": "devops", "event": "resume", "ts": "2026-03-02T10:00:00+00:00"}"#,
            r#"{"agent": "Explore", "ts": "2026-03-02T11:00:00+00:00"}"#,
        ]);
        let counts = tally(dir.path());

        assert_eq!(counts.get("devops"), Some(&1));
        assert_eq!(counts.get("Explore"), Some(&1));
        assert_eq!(generic_share(&counts), (1, 2));
    }

    #[test]
    fn a_line_that_is_not_a_row_is_skipped_rather_than_ending_the_file() {
        // The log is appended to by a hook: a truncated last line must not lose the rows
        // above it, and a blank line is the same non-row as a broken one.
        let dir = plane(&[
            r#"{"agent": "devops"}"#,
            "",
            "{not json",
            r#"["agent"]"#,
            r#"{"ts": "2026-03-02T09:33:00+00:00"}"#,
            r#"{"agent": "steward"}"#,
        ]);

        assert_eq!(tally(dir.path()).len(), 2);
    }

    #[test]
    fn a_plane_with_no_log_has_no_dispatches_rather_than_an_error() {
        let dir = tempfile::tempdir().unwrap();

        assert!(tally(dir.path()).is_empty());
        assert_eq!(generic_share(&tally(dir.path())), (0, 0));
    }

    #[test]
    fn a_dispatch_and_a_resume_are_logged_as_charter_logs_them() {
        let dir = tempfile::tempdir().unwrap();
        let when = chrono::DateTime::parse_from_rfc3339("2026-05-04T11:32:17+00:00")
            .unwrap()
            .with_timezone(&chrono::Utc);
        let p = record(dir.path(), " devops ", when, "box").unwrap();
        record_resume(dir.path(), "devops", when, "box").unwrap();
        assert_eq!(p, dir.path().join("personas/_dispatch/2026-05.box.jsonl"));
        assert_eq!(
            std::fs::read_to_string(&p).unwrap(),
            "{\"agent\": \"devops\", \"ts\": \"2026-05-04T11:32:17+00:00\"}\n\
             {\"agent\": \"devops\", \"event\": \"resume\", \"ts\": \"2026-05-04T11:32:17+00:00\"}\n"
        );
        assert_eq!(tally(dir.path()).get("devops"), Some(&1));
        assert_eq!(record(dir.path(), "  ", when, "box"), None);
    }

    #[test]
    fn the_host_is_one_safe_label() {
        assert_eq!(host_of("Aarons-MacBook.local"), "Aarons-MacBook");
        assert_eq!(host_of("bad host!"), "badhost");
        assert_eq!(host_of(""), "unknown");
        assert_eq!(host_of(&"x".repeat(40)).len(), 32);
    }
}
